pkg/security/security.go

```diff
@@ -351,12 +352,36 @@ type Caller struct {
 	Identities []string
 }
 
+type AuthContext struct {
+	// RequestContext is the context from request.
+	RequestContext context.Context
+	// Authenticators is the list of authenticators that were executed before in the auth chain.
+	Authenticators []string
+	// DelegatedAuthenticators are the list of delegated authenticators that main authenticators can
+	// choose to get identities from. For example, CidrAuthenticator can delegate the authentication to
+	// XfccAuthenticator once it validates the connection is from trusted cidr range. XfccAuthenticator
+	// can then extract identities from peer certificate.
+	DelegatedAuthenticators []Authenticator
+}
+
 type Authenticator interface {
-	Authenticate(ctx context.Context) (*Caller, error)
+	Authenticate(ctx AuthContext) (*Caller, error)
 	AuthenticatorType() string
 	AuthenticateRequest(req *http.Request) (*Caller, error)
 }
 
+func NewAuthContext(ctx context.Context) AuthContext {
```

Comment:
this should return a pointer to an `AuthContext` or else we will just copies around